File: 0401-0500/0441-arranging-coins.rs
struct Solution {}

impl Solution {
    pub fn arrange_coins(n: i32) -> i32 {
        let mut lo: i64 = 0;
        let mut hi: i64 = (n as i64) + 1;
        while lo < hi {
            let mid = lo + ((hi - lo) >> 1) as i64;
            if (1 + mid) * mid / 2 <= (n as i64) {
                lo = mid + 1;
            } else{
                hi = mid;
            }
        }
        (lo - 1) as i32
    }
}

fn main() {
    let n = 5;
    println!("{}", Solution::arrange_coins(n));
}

File: 0401-0500/0441-arranging-coins.cpp
class Solution {
public:
    int arrangeCoins(int n) {
        /*
            假设能构成完整K行
            (1 + K) * K / 2 <= n
            如何估计K的上下界？
            K * K / 2 < (1 + K) * K / 2 < (1 + K) * (1 + K) / 2
            K * K / 2 < n, K < sqrt(2 * n)
         */
         int lo = 0, hi = sqrt(2 * (long long)n) + 1;
         while (lo < hi) {
             int mid = lo + ((hi - lo) >> 1);
             (1 + (long long)mid) * (long long)mid / 2 <= n ? lo = mid + 1 : hi = mid;
         }
         return --lo;
    }
};
